### ISSUE TITLE
completion speed

### ISSUE THREAD

**Issue by A-Lovett** [nushell/nushell#11733]
### Question

Suppose you have a relatively slow completion function for an argument of a custom command. After you hit <tab> to bring up the completion menu, this function gets called after each individual key press, resulting in slow typing. In many cases, this is unnecessary because if the completion function doesn't depend on the word you are currently typing, you could simply cache the results from your prior call to the function.

Is there any way to address this issue in the current release of nushell? Or if it might require a patch to the source, might anyway, point me to a file to get started? My first idea was to cache the prior results from my completion function in $env. However, the results don't get saved to the top-level $env, even when I use the --env flag for my completion function. I'm guessing this is because whatever function calls the completion function doesn't have the --env flag, but I can't figure out what calls the completion function, or how I can get the completion function myself, given the function whose argument I am completing. Perhaps all of this is handled in rust and not nushell, but again I'd be curious about where in the source code.

Thanks.

**Comment by fdncred** [comment]
You'd need to make those changes in reedline probably since I think it's what handles the completion system. I wouldn't cache anything in the $env because environment space is not infinite, especially on Windows.

It's weird to hear you say this because I haven't heard anybody complain about the speed of completions for literally years. I'm wondering if you're using nushell on some decades old machine or something low powered like raspberry pi. Although, I haven't heard rasp users complain about completion speed either.

**Comment by A-Lovett** [comment]
Thanks for the reply. From what I've seen so far, it does seem like this would require a change in the source. To clarify a couple points:

1) If I cached results in $env, it would only be a single data structure, the most recent result from any cachable function, so it wouldn't accumulate over time. 

2) The slowness is from an external function. I'm using ros2, which manages a set of nodes for controlling a robot. Gathering a list of names for all current nodes takes 300-500ms, whereas other things can take longer. So if I have a custom command that takes a ros2 node and I want a completion for the node name, after I hit tab to bring up the completion menu, there's an extra 300-500ms delay after I type each letter.

**Comment by KAAtheWiseGit** [comment]
@fdncred I think that's because most completions are optimized for that.  Carapace, for example, doesn't suggest commit hashes in `git checkout`, as doing so would take too much time.

Caching could enable some nice opportunities.  For example, Alpine's package keeper `apk` has a subcommand `apk info <package>`.  But getting the list of packages via `apk list` takes almost a second.  So trying to implement completion for it without caching would be a fool's errand.

**Comment by fdncred** [comment]
> EDIT: I guess I could cache the results in a text file somewhere, along with a timestamp so the cache goes stale after ~10 seconds. Maybe that would be the simplest solution.

That could be the simplest solution indeed but you might also want to cache them in the in-memory sqlite database that you can access with the `stor` family of commands and `query db`.

**Comment by A-Lovett** [comment]
> 
> That could be the simplest solution indeed but you might also want to cache them in the in-memory sqlite database that you can access with the `stor` family of commands and `query db`.

Thanks for the suggestion. Caching to a text file in the nushell directory has worked well. I can definitely see the appeal of using a memory store in nushell, instead of touching the file system, but I can't find any documentation on how to use the stor commands on the nushell site, aside from the brief descriptions of each function. It's unclear to me how you would store some arbitrarily large data structure.

**Comment by StripedMonkey** [comment]
> It's weird to hear you say this because I haven't heard anybody complain about the speed of completions for literally years.

Your memory is at fault, I have made and been [complaining about this very issue](https://github.com/nushell/nu_scripts/issues/588) with regards to manpage completions in nu_scripts's `man` completions. This isn't an issue that has gone away.

Completions are absolutely a sticking point for me and likely others, but I suspect most either forgoe completions entirely or use "better" external mechanisms than try to depend on the integrated tools for completions. I find it hard to believe anyone is credibly using many completion mechanisms written in nushell because of the latency a "native" solution has by default.

**Comment by fdncred** [comment]
Oh, you're right. You did complain about 8 months ago. Thanks.

I use completions without issue. These are all the completions I use.
```nushell
# cargo completions
use "custom-completions/cargo/cargo-completions.nu" *
# use "custom-completions/winget/winget-completions.nu" *

# git/gh completions and aliases
use custom-completions/git/git-completions.nu *
use aliases/git/git-aliases.nu *
use custom-completions/gh/gh-completions.nu *

# broot completions
use '/Users/fdncred/.config/broot/launcher/nushell/br' *
```

**Comment by hadronized** [comment]
I’m having similar issues with external completions. I wrote a small wrapper around `kubectx` and `kubens`, and using the list of namespaces as source of completions makes everything slow down. I suspect the nu function passed as source of completions is at fault as it is invoked everytime the user types (and that function is really slow).

Is there a mitigation? I’m thinking of caching the result inside the completion function but not sure how to do that efficiently.

**Comment by StripedMonkey** [comment]
That is indeed the source of the slowdown, caching would work partially, but unfortunately getting it to a reasonable speed all the time needs more infrastructure than you'd hope. I believe there was a suggestion for using an in-memory db for some of that but invalidating that and regeneration of that cache basically needs to be done asynchronously at the current speed of at least the man completions I have been complaining about.

Most people, to my understanding, simply do not use the built-in completion for anything moderately complex, instead offloading the effort to something like carapace. That works wonderfully, but I wish nushell was a bit more powerful.

**Comment by hadronized** [comment]
> That is indeed the source of the slowdown, caching would work partially, but unfortunately getting it to a reasonable speed all the time needs more infrastructure than you'd hope. I believe there was a suggestion for using an in-memory db for some of that but invalidating that and regeneration of that cache basically needs to be done asynchronously at the current speed of at least the man completions I have been complaining about.
> 
> Most people, to my understanding, simply do not use the built-in completion for anything moderately complex, instead offloading the effort to something like carapace. That works wonderfully, but I wish nushell was a bit more powerful.

Thanks for the quick reply. I was thinking that the nu commands should probably simply be started in another process and piped via stdout/stdin to feed a trie or whatever structure is used for fuzzy search. Is that what you mentioned as being too much infrastructure for now?

**Comment by StripedMonkey** [comment]
I'm no nushell guru, so take what I'm saying with a grain of salt. Nushell doesn't use any kind of trie for completions natively. When you have built-in completions, you're just running a command that generates a record of possible completions + an optional description. There's exactly 0 optimization of that data structure, and there is no caching whatsoever. It just creates a list of possible results with whatever filtering you add to make "fuzzy" searches possible. There is no post processing by nushell itself to completions 

When I was referring to infrastructure, I'm saying that you would have to generate all completions, then store them somewhere between calls to the completion. The annoying part is that if you store it in-memory every time you have a new session, the list needs generating. If you store it on disk you are significantly slower but also you have to worry about regeneration of your completion cache when new completions come up. In the example of manpages that's relatively infrequent, but in the case of something more dynamic, that's dramatically more complex to determine on the fly.

The fact that there's no job control in nushell means that to manage this you have to depend on external tools that may not be present or portable across operating systems. It just becomes a lot to deal with. 

Or you could cop out and just use an external completion tool. Which is (unfortunately) the path of least resistance that I have taken. Carapace "just works" for almost everything I've really needed.

**Comment by hongxuchen** [comment]
Recently I found that on my machine fish completer performs *much* better than carapace completer in nushell. For example, carapace frequently hangs for quite a few seconds and later comes with *wrong candidates* when completing `apt` command; fish completer needs about 1-2s to complete in this case (meanwhile fish itself is still faster).

**Comment by KAAtheWiseGit** [comment]
Yeah, Nushell doesn't cache, so anything that needs expensive recalculations (network requests which carapace does sometimes, for example) will be slow.

I decided to use `stor` for my own completions:

```Nushell
def all-packages [] {
	try {
		stor open | query db "select * from __comp_apk" | length
	} catch {
		stor create --table-name __comp_apk --columns {name: str, version: str}
		get-all-packages | stor insert --table-name __comp_apk
	}

	stor open | query db "select * from __comp_apk" | get name
}
```

Where `get-all-packages` is a slow command.  One added benefit is that the cache persists between different completion invocations.

It's not optimal, however.  This cache doesn't handle updates, for example.  I feel like it'd be better to either have a more sophisticated cache implementation in `std` for use in completion scripts or a built-in completion caching mechanism.
